core/error.js

```diff
@@ -0,0 +1,231 @@
+      return result;
+    }
+    if (isMethodCall) {
+      const typeName = callSite.getTypeName();
+      const methodName = callSite.getMethodName();
+
+      if (functionName) {
+        if (typeName) {
+          if (!functionName.startsWith(typeName)) {
+            result += `${typeName}.`;
+          }
+        }
+        result += functionName;
+        if (methodName) {
+          if (!functionName.endsWith(methodName)) {
+            result += ` [as ${methodName}]`;
+          }
+        }
+      } else {
+        if (typeName) {
+          result += `${typeName}.`;
+        }
+        if (methodName) {
+          result += methodName;
+        } else {
+          result += "<anonymous>";
+        }
+      }
+    } else if (isConstructor) {
+      result += "new ";
+      if (functionName) {
+        result += functionName;
+      } else {
+        result += "<anonymous>";
+      }
+    } else if (functionName) {
+      result += functionName;
+    } else {
+      result += formatLocation(callSite);
+      return result;
+    }
+
+    result += ` (${formatLocation(callSite)})`;
+    return result;
+  }
+
+  function evaluateCallSite(callSite) {
+    return {
+      this: callSite.getThis(),
+      typeName: callSite.getTypeName(),
+      function: callSite.getFunction(),
+      functionName: callSite.getFunctionName(),
+      methodName: callSite.getMethodName(),
+      fileName: callSite.getFileName(),
+      lineNumber: callSite.getLineNumber(),
+      columnNumber: callSite.getColumnNumber(),
+      evalOrigin: callSite.getEvalOrigin(),
+      isToplevel: callSite.isToplevel(),
+      isEval: callSite.isEval(),
+      isNative: callSite.isNative(),
+      isConstructor: callSite.isConstructor(),
+      isAsync: callSite.isAsync(),
+      isPromiseAll: callSite.isPromiseAll(),
+      promiseIndex: callSite.getPromiseIndex(),
+    };
+  }
+
+  function createPrepareStackTrace(sourceMappingFn) {
+    return function prepareStackTrace(
+      error,
+      callSites,
+    ) {
+      const mappedCallSites = callSites.map(
+        (callSite) => {
+          const fileName = callSite.getFileName();
+          const lineNumber = callSite.getLineNumber();
+          const columnNumber = callSite.getColumnNumber();
+          if (
+            sourceMappingFn && fileName && lineNumber != null &&
+            columnNumber != null
+          ) {
+            return patchCallSite(
+              callSite,
+              sourceMappingFn({
+                fileName,
+                lineNumber,
+                columnNumber,
+              }),
+            );
+          }
+          return callSite;
+        },
+      );
+      Object.defineProperties(error, {
+        __callSiteEvals: { value: [], configurable: true },
+      });
+      const formattedCallSites = [];
+      for (const callSite of mappedCallSites) {
+        error.__callSiteEvals.push(Object.freeze(evaluateCallSite(callSite)));
```

Comment:
I'm not sure if it's still the case but `Object.freeze()` used to carry a pretty big performance penalty, both to execute and because of the knock-on effects of changing the object's hidden class.

A quick look at V8's source suggests it's still a complicated affair (`JSObject::PreventExtensionsWithTransition()` in `src/objects/js-objects.cc`) so I'd try to avoid it if possible.

Reply:
Yeah I'm not sure if it makes sense to freeze it as those frames are immediately captured by `deno_core::JsError`

Reply:
Not necessarily immediately, this function is called when `error.stack` is first accessed by the user. `JsError` is created on termination.

Nevertheless, as of #5254 the `__callSiteEvals` field is `configurable: true`, so there's definitely no point in freezing the value. It's a holdover.

Reply:
Good point, I'll remove the freezing then.